### ISSUE TITLE
New failure with regex match including non-ASCII characters

### ISSUE THREAD

**Issue by nalimilan** [JuliaLang/julia#24157]
This recently appeared on master. I'm afraid we introduced this in https://github.com/JuliaLang/julia/pull/22511. Cc: @bkamins 
```julia
julia> matchall(r"fé", "café")
ERROR: BoundsError: attempt to access "café"
  at index [5]
Stacktrace:
 [1] SubString{String}(::String, ::Int64, ::Int64) at ./strings/types.jl:33
 [2] matchall(::Regex, ::String, ::Bool) at ./regex.jl:272
 [3] matchall(::Regex, ::String) at ./regex.jl:252
```

**Comment by bkamins** [comment]
I believe the fix of this is already submitted in #24116, and in particular
https://github.com/JuliaLang/julia/pull/24116/files#diff-7422507bb03a8ecee35360dc9380b9caR269
and two additional fixed in the same PR of the same kind.
Today I should push a revision of that PR following your comments. I will add the above to test cases.

**Comment by vtjnash** [comment]
This regression was fixed prior to the relevant release, a long time ago:

```
julia> eachmatch(r"fé", "café") |> collect
1-element Array{RegexMatch,1}:
 RegexMatch("fé")
```
